File: customer-app/lib/models/order.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'order.freezed.dart';
part 'order.g.dart';

/// Order status enum
enum OrderStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('confirmed')
  confirmed,
  @JsonValue('preparing')
  preparing,
  @JsonValue('ready')
  ready,
  @JsonValue('picked_up')
  pickedUp,
  @JsonValue('on_the_way')
  onTheWay,
  @JsonValue('delivered')
  delivered,
  @JsonValue('cancelled')
  cancelled,
}

/// Payment method enum
enum OrderPaymentMethod {
  @JsonValue('cash')
  cash,
  @JsonValue('card')
  card,
  @JsonValue('wallet')
  wallet,
}

/// Payment status enum
enum PaymentStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('paid')
  paid,
  @JsonValue('failed')
  failed,
  @JsonValue('refunded')
  refunded,
}

/// Order item model
@freezed
class OrderItem with _$OrderItem {
  const factory OrderItem({
    required String menuItemId,
    required String name,
    String? nameAr,
    String? image,
    required double price,
    required int quantity,
    @Default([]) List<OrderItemAddon> addons,
    @Default([]) List<OrderItemVariation> variations,
    String? specialInstructions,
  }) = _OrderItem;

  factory OrderItem.fromJson(Map<String, dynamic> json) =>
      _$OrderItemFromJson(json);

  const OrderItem._();

  String get displayName => nameAr ?? name;
  double get total => (price + addonsTotal + variationsTotal) * quantity;
  double get addonsTotal =>
      addons.fold(0.0, (sum, a) => sum + (a.price * a.quantity));
  double get variationsTotal =>
      variations.fold(0.0, (sum, v) => sum + v.price);
}

/// Order item addon
@freezed
class OrderItemAddon with _$OrderItemAddon {
  const factory OrderItemAddon({
    required String name,
    String? nameAr,
    required double price,
    @Default(1) int quantity,
  }) = _OrderItemAddon;

  factory OrderItemAddon.fromJson(Map<String, dynamic> json) =>
      _$OrderItemAddonFromJson(json);
}

/// Order item variation
@freezed
class OrderItemVariation with _$OrderItemVariation {
  const factory OrderItemVariation({
    required String name,
    String? nameAr,
    required String option,
    String? optionAr,
    @Default(0.0) double price,
  }) = _OrderItemVariation;

  factory OrderItemVariation.fromJson(Map<String, dynamic> json) =>
      _$OrderItemVariationFromJson(json);
}

/// Delivery address in order
@freezed
class OrderDeliveryAddress with _$OrderDeliveryAddress {
  const factory OrderDeliveryAddress({
    required String name,
    required String address,
    required String area,
    required String city,
    String? building,
    String? floor,
    String? apartment,
    String? landmark,
    required List<double> coordinates,
  }) = _OrderDeliveryAddress;

  factory OrderDeliveryAddress.fromJson(Map<String, dynamic> json) =>
      _$OrderDeliveryAddressFromJson(json);

  const OrderDeliveryAddress._();

  double get latitude => coordinates.length > 1 ? coordinates[1] : 0;
  double get longitude => coordinates.isNotEmpty ? coordinates[0] : 0;

  String get fullAddress => '$address، $area، $city';
}

/// Driver info in order
@freezed
class OrderDriver with _$OrderDriver {
  const factory OrderDriver({
    required String id,
    required String name,
    String? phone,
    String? avatar,
    double? rating,
    List<double>? currentLocation,
  }) = _OrderDriver;

  factory OrderDriver.fromJson(Map<String, dynamic> json) =>
      _$OrderDriverFromJson(json);

  const OrderDriver._();

  double? get latitude =>
      currentLocation != null && currentLocation!.length > 1
          ? currentLocation![1]
          : null;
  double? get longitude =>
      currentLocation != null && currentLocation!.isNotEmpty
          ? currentLocation![0]
          : null;
}

/// Restaurant info in order
@freezed
class OrderRestaurant with _$OrderRestaurant {
  const factory OrderRestaurant({
    required String id,
    required String name,
    String? nameAr,
    String? logo,
    String? phone,
    List<double>? location,
  }) = _OrderRestaurant;

  factory OrderRestaurant.fromJson(Map<String, dynamic> json) =>
      _$OrderRestaurantFromJson(json);

  const OrderRestaurant._();

  String get displayName => nameAr ?? name;
}

/// Status history entry
@freezed
class OrderStatusHistory with _$OrderStatusHistory {
  const factory OrderStatusHistory({
    required OrderStatus status,
    required DateTime timestamp,
    String? note,
  }) = _OrderStatusHistory;

  factory OrderStatusHistory.fromJson(Map<String, dynamic> json) =>
      _$OrderStatusHistoryFromJson(json);
}

/// Main Order model
@freezed
class Order with _$Order {
  const factory Order({
    @JsonKey(name: '_id') required String id,
    required String orderNumber,
    required OrderRestaurant restaurant,
    required List<OrderItem> items,
    required OrderDeliveryAddress deliveryAddress,
    required OrderStatus status,
    required OrderPaymentMethod paymentMethod,
    required PaymentStatus paymentStatus,
    required double subtotal,
    required double deliveryFee,
    @Default(0.0) double discount,
    required double total,
    OrderDriver? driver,
    String? notes,
    String? cancellationReason,
    DateTime? estimatedDeliveryTime,
    @Default([]) List<OrderStatusHistory> statusHistory,
    required DateTime createdAt,
    DateTime? updatedAt,
  }) = _Order;

  factory Order.fromJson(Map<String, dynamic> json) => _$OrderFromJson(json);

  const Order._();

  /// Get status display name in Arabic
  String get statusDisplayName {
    switch (status) {
      case OrderStatus.pending:
        return 'في انتظار التأكيد';
      case OrderStatus.confirmed:
        return 'تم التأكيد';
      case OrderStatus.preparing:
        return 'جاري التحضير';
      case OrderStatus.ready:
        return 'جاهز للاستلام';
      case OrderStatus.pickedUp:
        return 'تم الاستلام';
      case OrderStatus.onTheWay:
        return 'في الطريق إليك';
      case OrderStatus.delivered:
        return 'تم التوصيل';
      case OrderStatus.cancelled:
        return 'ملغي';
    }
  }

  /// Check if order is active (not completed or cancelled)
  bool get isActive =>
      status != OrderStatus.delivered && status != OrderStatus.cancelled;

  /// Check if order can be cancelled
  bool get canCancel =>
      status == OrderStatus.pending || status == OrderStatus.confirmed;

  /// Get progress percentage (0-100)
  int get progressPercentage {
    switch (status) {
      case OrderStatus.pending:
        return 10;
      case OrderStatus.confirmed:
        return 25;
      case OrderStatus.preparing:
        return 40;
      case OrderStatus.ready:
        return 55;
      case OrderStatus.pickedUp:
        return 70;
      case OrderStatus.onTheWay:
        return 85;
      case OrderStatus.delivered:
        return 100;
      case OrderStatus.cancelled:
        return 0;
    }
  }

  /// Get total items count
  int get totalItems => items.fold(0, (sum, item) => sum + item.quantity);
}

File: backend/src/models/Order.ts
import mongoose, { Schema, Document, Types } from 'mongoose';
import dayjs from 'dayjs';
import { ILocation, OrderStatus, PaymentStatus, PaymentMethod } from '../types';

export interface IOrderItemAddon {
  addonId: Types.ObjectId;
  name: string;
  nameAr: string;
  price: number;
  quantity: number;
}

export interface IOrderItemVariation {
  variationId: Types.ObjectId;
  variationName: string;
  optionId: Types.ObjectId;
  optionName: string;
  optionNameAr: string;
  price: number;
}

export interface IOrderItem {
  _id: Types.ObjectId;
  menuItemId: Types.ObjectId;
  name: string;
  nameAr: string;
  image?: string;
  basePrice: number;
  quantity: number;
  selectedAddons: IOrderItemAddon[];
  selectedVariations: IOrderItemVariation[];
  specialInstructions?: string;
  itemTotal: number;
}

export interface IDeliveryAddress {
  name: string;
  address: string;
  area: string;
  building?: string;
  floor?: string;
  apartment?: string;
  landmark?: string;
  phone: string;
  location: ILocation;
}

export interface IStatusHistory {
  status: OrderStatus;
  timestamp: Date;
  note?: string;
  updatedBy?: Types.ObjectId;
}

export interface IOrderRating {
  restaurant?: number;
  driver?: number;
  food?: number;
  comment?: string;
  ratedAt?: Date;
}

export interface IOrder extends Document {
  _id: Types.ObjectId;
  orderNumber: string;

  // Parties
  customerId: Types.ObjectId;
  restaurantId: Types.ObjectId;
  driverId?: Types.ObjectId;

  // Items
  items: IOrderItem[];

  // Pricing
  subtotal: number;
  deliveryFee: number;
  serviceFee: number;
  discount: number;
  total: number;

  // Coupon
  couponId?: Types.ObjectId;
  couponCode?: string;
  couponDiscount: number;

  // Delivery Address
  deliveryAddress: IDeliveryAddress;

  // Notes
  customerNotes?: string;
  restaurantNotes?: string;

  // Status
  status: OrderStatus;
  statusHistory: IStatusHistory[];

  // Cancellation
  isCancelled: boolean;
  cancelReason?: string;
  cancelledBy?: 'customer' | 'restaurant' | 'driver' | 'admin';
  cancelledAt?: Date;

  // Payment
  paymentMethod: PaymentMethod;
  paymentStatus: PaymentStatus;
  paymentId?: string;
  transactionId?: Types.ObjectId;
  paidAt?: Date;

  // Timing
  estimatedDeliveryTime?: Date;
  restaurantAcceptedAt?: Date;
  preparationStartedAt?: Date;
  readyAt?: Date;
  driverAssignedAt?: Date;
  pickedUpAt?: Date;
  deliveredAt?: Date;

  // Rating
  rating?: IOrderRating;

  // Financials
  restaurantEarnings: number;
  driverEarnings: number;
  platformEarnings: number;

  // Flags
  isScheduled: boolean;
  scheduledFor?: Date;
  isReorder: boolean;
  originalOrderId?: Types.ObjectId;

  createdAt: Date;
  updatedAt: Date;
}

const orderItemAddonSchema = new Schema<IOrderItemAddon>(
  {
    addonId: { type: Schema.Types.ObjectId, required: true },
    name: { type: String, required: true },
    nameAr: { type: String, required: true },
    price: { type: Number, required: true },
    quantity: { type: Number, required: true, min: 1 },
  },
  { _id: false }
);

const orderItemVariationSchema = new Schema<IOrderItemVariation>(
  {
    variationId: { type: Schema.Types.ObjectId, required: true },
    variationName: { type: String, required: true },
    optionId: { type: Schema.Types.ObjectId, required: true },
    optionName: { type: String, required: true },
    optionNameAr: { type: String, required: true },
    price: { type: Number, required: true },
  },
  { _id: false }
);

const orderItemSchema = new Schema<IOrderItem>(
  {
    menuItemId: {
      type: Schema.Types.ObjectId,
      ref: 'MenuItem',
      required: true,
    },
    name: { type: String, required: true },
    nameAr: { type: String, required: true },
    image: { type: String },
    basePrice: { type: Number, required: true },
    quantity: { type: Number, required: true, min: 1 },
    selectedAddons: { type: [orderItemAddonSchema], default: [] },
    selectedVariations: { type: [orderItemVariationSchema], default: [] },
    specialInstructions: { type: String },
    itemTotal: { type: Number, required: true },
  },
  { _id: true }
);

const deliveryAddressSchema = new Schema<IDeliveryAddress>(
  {
    name: { type: String, required: true },
    address: { type: String, required: true },
    area: { type: String, required: true },
    building: { type: String },
    floor: { type: String },
    apartment: { type: String },
    landmark: { type: String },
    phone: { type: String, required: true },
    location: {
      type: { type: String, enum: ['Point'], default: 'Point' },
      coordinates: { type: [Number], required: true },
    },
  },
  { _id: false }
);

const statusHistorySchema = new Schema<IStatusHistory>(
  {
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'preparing', 'ready', 'picked_up', 'on_the_way', 'delivered', 'cancelled'],
      required: true,
    },
    timestamp: { type: Date, default: Date.now },
    note: { type: String },
    updatedBy: { type: Schema.Types.ObjectId, ref: 'User' },
  },
  { _id: false }
);

const orderRatingSchema = new Schema<IOrderRating>(
  {
    restaurant: { type: Number, min: 1, max: 5 },
    driver: { type: Number, min: 1, max: 5 },
    food: { type: Number, min: 1, max: 5 },
    comment: { type: String },
    ratedAt: { type: Date },
  },
  { _id: false }
);

const orderSchema = new Schema<IOrder>(
  {
    orderNumber: {
      type: String,
      unique: true,
    },

    // Parties
    customerId: {
      type: Schema.Types.ObjectId,
      ref: 'Customer',
      required: true,
    },
    restaurantId: {
      type: Schema.Types.ObjectId,
      ref: 'Restaurant',
      required: true,
    },
    driverId: {
      type: Schema.Types.ObjectId,
      ref: 'Driver',
    },

    // Items
    items: {
      type: [orderItemSchema],
      required: true,
      validate: {
        validator: function (items: IOrderItem[]) {
          return items.length > 0;
        },
        message: 'Order must have at least one item',
      },
    },

    // Pricing
    subtotal: {
      type: Number,
      required: true,
      min: 0,
    },
    deliveryFee: {
      type: Number,
      default: 0,
      min: 0,
    },
    serviceFee: {
      type: Number,
      default: 0,
      min: 0,
    },
    discount: {
      type: Number,
      default: 0,
      min: 0,
    },
    total: {
      type: Number,
      required: true,
      min: 0,
    },

    // Coupon
    couponId: {
      type: Schema.Types.ObjectId,
      ref: 'Coupon',
    },
    couponCode: {
      type: String,
    },
    couponDiscount: {
      type: Number,
      default: 0,
      min: 0,
    },

    // Delivery Address
    deliveryAddress: {
      type: deliveryAddressSchema,
      required: true,
    },

    // Notes
    customerNotes: {
      type: String,
    },
    restaurantNotes: {
      type: String,
    },

    // Status
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'preparing', 'ready', 'picked_up', 'on_the_way', 'delivered', 'cancelled'],
      default: 'pending',
    },
    statusHistory: {
      type: [statusHistorySchema],
      default: [],
    },

    // Cancellation
    isCancelled: {
      type: Boolean,
      default: false,
    },
    cancelReason: {
      type: String,
    },
    cancelledBy: {
      type: String,
      enum: ['customer', 'restaurant', 'driver', 'admin'],
    },
    cancelledAt: {
      type: Date,
    },

    // Payment
    paymentMethod: {
      type: String,
      enum: ['cash', 'card', 'wallet'],
      default: 'cash',
    },
    paymentStatus: {
      type: String,
      enum: ['pending', 'paid', 'failed', 'refunded'],
      default: 'pending',
    },
    paymentId: {
      type: String,
    },
    transactionId: {
      type: Schema.Types.ObjectId,
      ref: 'Transaction',
    },
    paidAt: {
      type: Date,
    },

    // Timing
    estimatedDeliveryTime: {
      type: Date,
    },
    restaurantAcceptedAt: {
      type: Date,
    },
    preparationStartedAt: {
      type: Date,
    },
    readyAt: {
      type: Date,
    },
    driverAssignedAt: {
      type: Date,
    },
    pickedUpAt: {
      type: Date,
    },
    deliveredAt: {
      type: Date,
    },

    // Rating
    rating: {
      type: orderRatingSchema,
    },

    // Financials
    restaurantEarnings: {
      type: Number,
      default: 0,
      min: 0,
    },
    driverEarnings: {
      type: Number,
      default: 0,
      min: 0,
    },
    platformEarnings: {
      type: Number,
      default: 0,
      min: 0,
    },

    // Flags
    isScheduled: {
      type: Boolean,
      default: false,
    },
    scheduledFor: {
      type: Date,
    },
    isReorder: {
      type: Boolean,
      default: false,
    },
    originalOrderId: {
      type: Schema.Types.ObjectId,
      ref: 'Order',
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Indexes
orderSchema.index({ orderNumber: 1 }, { unique: true });
orderSchema.index({ customerId: 1 });
orderSchema.index({ restaurantId: 1 });
orderSchema.index({ driverId: 1 });
orderSchema.index({ status: 1 });
orderSchema.index({ createdAt: -1 });
orderSchema.index({ restaurantId: 1, status: 1 });
orderSchema.index({ driverId: 1, status: 1 });

// Pre-save hook to generate order number
orderSchema.pre('save', async function () {
  if (!this.orderNumber) {
    const date = dayjs().format('YYMMDD');
    const count = await mongoose.model('Order').countDocuments({
      createdAt: {
        $gte: dayjs().startOf('day').toDate(),
        $lte: dayjs().endOf('day').toDate(),
      },
    });
    this.orderNumber = `BAG-${date}-${(count + 1).toString().padStart(4, '0')}`;
  }

  // Add initial status to history
  if ((this as any).isNew && this.statusHistory.length === 0) {
    this.statusHistory.push({
      status: this.status,
      timestamp: new Date(),
    });
  }
});

// Virtual for customer details
orderSchema.virtual('customer', {
  ref: 'Customer',
  localField: 'customerId',
  foreignField: '_id',
  justOne: true,
});

// Virtual for restaurant details
orderSchema.virtual('restaurant', {
  ref: 'Restaurant',
  localField: 'restaurantId',
  foreignField: '_id',
  justOne: true,
});

// Virtual for driver details
orderSchema.virtual('driver', {
  ref: 'Driver',
  localField: 'driverId',
  foreignField: '_id',
  justOne: true,
});

export const Order = mongoose.model<IOrder>('Order', orderSchema);
export default Order;
